pkg/networkservice/common/heal/server.go

```diff
@@ -0,0 +1,436 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package heal provides a chain element that carries out proper nsm healing from client to endpoint
+package heal
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	"github.com/edwarnicke/serialize"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/discover"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/addressof"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+)
+
+type ctxWrapper struct {
+	ctx    context.Context
+	cancel func()
+}
+
+type clientConnInfo struct {
+	ctx context.Context
+	cc  networkservice.MonitorConnectionClient
+}
+
+type connection struct {
+	*networkservice.Connection
+	ctx context.Context
+}
+
+type healServer struct {
+	ctx                   context.Context
+	clients               clientConnMap
+	onHeal                *networkservice.NetworkServiceClient
+	cancelHealMap         map[string]*ctxWrapper
+	cancelHealMapExecutor serialize.Executor
+	conns                 connectionMap
+}
+
+// NewServer - creates a new networkservice.NetworkServiceServer chain element that implements the healing algorithm
+//             - ctx    - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
+//             - client - networkservice.MonitorConnectionClient that can be used to call MonitorConnection against the endpoint
+//             - onHeal - *networkservice.NetworkServiceClient.  Since networkservice.NetworkServiceClient is an interface
+//                        (and thus a pointer) *networkservice.NetworkServiceClient is a double pointer.  Meaning it
+//                        points to a place that points to a place that implements networkservice.NetworkServiceClient
+//                        This is done because when we use heal.NewClient as part of a chain, we may not *have*
+//                        a pointer to this
+//                        client used 'onHeal'.  If we detect we need to heal, onHeal.Request is used to heal.
+//                        If onHeal is nil, then we simply set onHeal to this client chain element
+//                        If we are part of a larger chain or a server, we should pass the resulting chain into
+//                        this constructor before we actually have a pointer to it.
+//                        If onHeal nil, onHeal will be pointed to the returned networkservice.NetworkServiceClient
+func NewServer(ctx context.Context, onHeal *networkservice.NetworkServiceClient) (networkservice.NetworkServiceServer, RegisterClientFunc) {
```

Comment:
Please let's do not return two parameters from `NewServer`. It makes `heal` not consistent with other chain elements.

Reply:
Done